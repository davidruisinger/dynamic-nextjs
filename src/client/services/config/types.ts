export type Config = {
  url_product_base: string
  url_template_base_web: string
  url_template_base_ios: string
  url_template_base_android: string
}
