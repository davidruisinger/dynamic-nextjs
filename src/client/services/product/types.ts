export type Product = {
  id: string
  shop: string
  name_en: string
  desc_en: string
  images: ProductImage[]
  price_eur_regular: number
  price_eur_discounted: number
  variations: ProductVariation[]
}

export type ProductImage = {
  url: string
}

export type ProductVariation = {
  id: string
  name: string
}
