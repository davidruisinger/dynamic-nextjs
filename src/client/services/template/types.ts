export type TemplateMeta = {
  app_name: string
  app_logo: string
}

export type ProductListElement = {
  type: 'ProductList'
  content: string[]
}

export type MarkdownElement = {
  type: 'Markdown'
  content: string
}

export type TemplatePageContent = ProductListElement | MarkdownElement

export type TemplatePage = {
  href: string
  label: string
  content: TemplatePageContent[]
}

export type TemplateVariables = {
  color_canvas_ground: string
  color_canvas_first: string
  color_canvas_second: string
  color_canvas_third: string

  color_nav_bar_item: string
  color_nav_bar_item_active: string

  color_text_default: string
  color_text_highlight: string
  color_text_semi: string

  color_action_default: string
  color_action_primary: string
  color_action_secondary: string
  color_action_label: string

  color_success: string
  color_danger: string
}

export type Template = {
  id: string
  meta: TemplateMeta
  pages: TemplatePage[]
  variables: TemplateVariables
}
