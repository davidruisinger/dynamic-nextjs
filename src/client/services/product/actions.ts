import { createAction } from 'typesafe-actions'
import { Product } from './types'

/**
 * Redux action to fetch multiple products by ID.
 * @param {Object} config
 * @param {string[]} config.productIds - The productIds to fetch.
 */
export const fetchProducts = createAction('GET_PRODUCTS_REQUEST', resolve => {
  return ({ productIds }: { productIds: string[] }) => resolve({ productIds })
})

/**
 * Redux action to process an array of fetched products.
 * @param {Object} result
 * @param {string} result.products - The fetched products.
 */
export const fetchProductsSuccess = createAction(
  'GET_PRODUCTS_SUCCESS',
  resolve => {
    return ({ products }: { products: Product[] }) => resolve({ products })
  }
)

/**
 * Redux action to process an error while fetching products.
 * @param {Object} error
 * @param {string} error.message - The error message.
 */
export const fetchProductsFail = createAction('GET_PRODUCTS_FAIL', resolve => {
  return ({ message }: { message: string }) => resolve({ message })
})
