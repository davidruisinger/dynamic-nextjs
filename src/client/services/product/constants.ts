export const SERVICE_NAME = 'product_service'
