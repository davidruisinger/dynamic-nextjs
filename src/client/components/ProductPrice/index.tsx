import React, { FunctionComponent } from 'react'
import styled from '../_utils/styledComponents'
import { Product } from '../../services/product/types'

interface Props {
  product: Product
}

const toCurrencyString = (price: number) => `â‚¬ ${price / 100}`

const OriginalPrice = styled.span`
  color: ${props => props.theme.color_text_semi};
  text-decoration: line-through ${props => props.theme.color_text_default};
`

const SellingPrice = styled.span`
  margin: 10px;
  color: ${props => props.theme.color_text_highlight};
`

const ProductPrice: FunctionComponent<Props> = ({ product }) => {
  const discountedPrice = product.price_eur_discounted
  const regularPrice = product.price_eur_regular
  const sellingPrice = Math.min(regularPrice, discountedPrice)

  return (
    <span>
      {sellingPrice < regularPrice ? (
        <>
          <OriginalPrice>{toCurrencyString(regularPrice)}</OriginalPrice>{' '}
        </>
      ) : null}
      <SellingPrice>{toCurrencyString(sellingPrice)}</SellingPrice>
    </span>
  )
}

export default ProductPrice
