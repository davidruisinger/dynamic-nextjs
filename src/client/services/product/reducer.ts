import { ActionType, getType } from 'typesafe-actions'
import * as actions from './actions'
import { Product } from './types'

export type ProductState = Readonly<{
  products: Product[]
  isFetching: boolean
  error: string
}>

const initialState: ProductState = {
  products: [],
  isFetching: false,
  error: null,
}

export default (
  state: ProductState = initialState,
  action: ActionType<typeof actions>
): ProductState => {
  switch (action.type) {
    case getType(actions.fetchProducts):
      return {
        ...state,
        isFetching: true,
        error: null,
      }

    case getType(actions.fetchProductsSuccess):
      return {
        ...state,
        products: [...state.products, ...action.payload.products],
        isFetching: false,
      }

    case getType(actions.fetchProductsFail):
      return {
        ...state,
        isFetching: false,
        error: action.payload.message,
      }

    default:
      return state
  }
}
