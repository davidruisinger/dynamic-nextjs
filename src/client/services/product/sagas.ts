import { getType } from 'typesafe-actions'
import { select, put, takeLatest, call } from 'redux-saga/effects'
import { Product } from './types'
import * as actions from './actions'
import { getProducts } from './selectors'
import { fetchConfigSaga } from '../config/sagas'
import { getShopId, getUrlProductBase } from '../config/selectors'

export function* fetchProductsSaga(
  action: ReturnType<typeof actions.fetchProducts>
) {
  try {
    // Ensure that the config is already loaded...
    yield call(fetchConfigSaga)
    // ... get the shopId & base URL...
    const shopId = yield select(getShopId)
    if (!shopId) throw new Error('shopId missing')
    const baseUrl = yield select(getUrlProductBase)
    if (!baseUrl) throw new Error('baseUrl missing')
    // ..and start fetching the template
    const { productIds } = action.payload
    // Get the existing products from the store and filter those products out
    const existingProducts = yield select(getProducts, { productIds })
    const existingProductIds = existingProducts.map(
      (product: Product) => product.id
    )
    const filteredProductIds = productIds.filter(
      id => !existingProductIds.includes(id)
    )
    const response: Response = yield call(
      fetch,
      `${baseUrl}?shop=${shopId}&id=${filteredProductIds.join('&id=')}`
    )
    const json: Product[] = yield call([response, response.json])
    if (!response.ok) {
      throw new Error(response.statusText)
    }
    yield put(actions.fetchProductsSuccess({ products: json }))
  } catch (e) {
    yield put(
      actions.fetchProductsFail({
        message: e.message || 'Unknown error',
      })
    )
  }
}

export default [takeLatest(getType(actions.fetchProducts), fetchProductsSaga)]
