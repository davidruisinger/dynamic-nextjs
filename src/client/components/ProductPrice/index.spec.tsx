import 'jest-styled-components'
import React from 'react'
import testRenderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import ProductPrice from './index'
import { templateMock } from '../../services/template/__tests__/_mocks'
import { productsMock } from '../../services/product/__tests__/_mocks'
import { Product } from '../../services/product/types'

const productWithoutDiscount: Product = {
  ...productsMock[0],
  price_eur_regular: 100,
  price_eur_discounted: 100,
}
const productWithDiscount: Product = {
  ...productWithoutDiscount,
  price_eur_discounted: 90,
}

const regularPriceTree = testRenderer
  .create(
    <ThemeProvider theme={templateMock.variables}>
      <ProductPrice product={productWithoutDiscount} />
    </ThemeProvider>
  )
  .toJSON()
const discountedPriceTree = testRenderer
  .create(
    <ThemeProvider theme={templateMock.variables}>
      <ProductPrice product={productWithDiscount} />
    </ThemeProvider>
  )
  .toJSON()

describe('<ProductPrice />', () => {
  it('matches the snapshot', () => {
    expect(regularPriceTree).toMatchSnapshot()
    expect(discountedPriceTree).toMatchSnapshot()
  })

  it('renders a single price if there is no discount', () => {
    expect(regularPriceTree.children).toHaveLength(1)
    expect(regularPriceTree.children[0].children[0]).toEqual('€ 1')
  })

  it('renders the regular and the discounted price if there is a discount', () => {
    // 3 child nodes because we have a node with a ' ' (space character) inbetween
    expect(discountedPriceTree.children).toHaveLength(3)
    expect(discountedPriceTree.children[0].children[0]).toEqual('€ 1')
    expect(discountedPriceTree.children[1]).toEqual(' ')
    expect(discountedPriceTree.children[2].children[0]).toEqual('€ 0.9')
  })

  it('renders with the correct colors', () => {
    // 3 child nodes because we have a node with a ' ' (space character) inbetween
    expect(regularPriceTree.children[0]).toHaveStyleRule(
      'color',
      templateMock.variables.color_text_highlight
    )
    expect(discountedPriceTree.children[0]).toHaveStyleRule(
      'color',
      templateMock.variables.color_text_semi
    )
    expect(discountedPriceTree.children[2]).toHaveStyleRule(
      'color',
      templateMock.variables.color_text_highlight
    )
  })
})
