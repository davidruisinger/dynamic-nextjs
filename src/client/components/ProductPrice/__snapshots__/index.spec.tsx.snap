// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPrice /> matches the snapshot 1`] = `
.c0 {
  margin: 10px;
  color: #d8833d;
}

<span>
  <span
    className="c0"
  >
    € 1
  </span>
</span>
`;

exports[`<ProductPrice /> matches the snapshot 2`] = `
.c0 {
  color: #9b9b9b;
  -webkit-text-decoration: line-through #2f464f;
  text-decoration: line-through #2f464f;
}

.c1 {
  margin: 10px;
  color: #d8833d;
}

<span>
  <span
    className="c0"
  >
    € 1
  </span>
   
  <span
    className="c1"
  >
    € 0.9
  </span>
</span>
`;
